projects/lights_control/src/lights_can.c

```diff
@@ -0,0 +1,89 @@
+#include "can.h"
+#include "can_hw.h"
+#include "can_msg_defs.h"
+#include "can_unpack.h"
+#include "event_queue.h"
+#include "status.h"
+
+#include "exported_enums.h"
+#include "lights_can.h"
+#include "lights_events.h"
+
+// RX handler for signal type events.
+static StatusCode prv_rx_signal_handler(const CANMessage *msg, void *context,
+                                        CANAckStatus *ack_reply) {
+  uint8_t signal_type = 0;
+  uint8_t state = 0;
+  status_ok_or_return(CAN_UNPACK_LIGHTS_SIGNAL_TYPES(msg, &signal_type, &state));
+  if (signal_type >= NUM_LIGHTS_SIGNAL_TYPES) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "Unsupported signals type.");
+  }
+  return event_raise((state) ? LIGHTS_EVENT_SIGNAL_ON : LIGHTS_EVENT_SIGNAL_OFF,
+                     (LightsEventSignalMode)signal_type);
+}
+
+// RX handler for generic lights.
+static StatusCode prv_rx_generic_lights_handler(const CANMessage *msg, void *context,
+                                                CANAckStatus *ack_reply) {
+  LightsCanSettings *settings = (LightsCanSettings *)context;
+  uint8_t light_type = 0;
```

Comment:
Seems like a lot of duplicated code

Reply:
I can't think of a clean way of reducing it :( too many things change in every function.